File: backend/src/bond_monitor/application/notifications/consumer.py
"""Background Redis consumer for in-app notifications."""

from __future__ import annotations

import asyncio
import logging

from bond_monitor.infrastructure.notifications.notifications_repository import NotificationsRepository
from bond_monitor.infrastructure.notifications.redis_bus import NotificationBus
from bond_monitor.infrastructure.persistence.database import get_session_factory

logger = logging.getLogger(__name__)


class NotificationConsumer:
    def __init__(self, redis_url: str, *, consumer_name: str = "api-1") -> None:
        self._bus = NotificationBus(redis_url)
        self._consumer_name = consumer_name
        self._task: asyncio.Task | None = None
        self._stop = asyncio.Event()

    async def start(self) -> None:
        try:
            self._bus.ensure_consumer_group()
        except Exception:
            logger.warning("Redis consumer group setup failed", exc_info=True)
            return
        self._stop.clear()
        self._task = asyncio.create_task(self._loop())
        logger.info("Notification consumer started")

    async def stop(self) -> None:
        self._stop.set()
        if self._task is not None:
            await self._task

    async def _loop(self) -> None:
        session_factory = get_session_factory()
        while not self._stop.is_set():
            try:
                messages = await asyncio.to_thread(
                    self._bus.read_group,
                    self._consumer_name,
                    count=20,
                )
                if not messages:
                    continue
                async with session_factory() as session:
                    repo = NotificationsRepository(session)
                    for message in messages:
                        await repo.upsert_from_bus(
                            fingerprint=message.fingerprint,
                            portfolio_id=message.portfolio_id,
                            kind=message.kind,
                            payload=message.payload,
                            urgency=message.urgency,
                        )
                        await asyncio.to_thread(self._bus.ack, message.message_id)
            except Exception:
                logger.exception("Notification consumer iteration failed")
                await asyncio.sleep(2)

File: backend/internal/application/notifications/consumer.go
package notifications

import (
	"context"
	"log/slog"
	"time"

	"github.com/tonatos/instrumenta/backend/internal/application"
	"github.com/tonatos/instrumenta/backend/internal/infrastructure/notifications"
	"github.com/tonatos/instrumenta/backend/internal/infrastructure/persistence"
)

// Consumer reads Redis notification stream into SQLite read-model.
type Consumer struct {
	bus    *notifications.RedisBus
	repo   *persistence.NotificationsRepository
	name   string
	stop   chan struct{}
	done   chan struct{}
	logger *slog.Logger
}

func NewConsumer(redisURL string, repo *persistence.NotificationsRepository, logger *slog.Logger) *Consumer {
	return &Consumer{
		bus:    notifications.NewRedisBus(redisURL),
		repo:   repo,
		name:   "api-1",
		stop:   make(chan struct{}),
		done:   make(chan struct{}),
		logger: logger,
	}
}

func (c *Consumer) Start(ctx context.Context) error {
	if c.bus == nil || c.repo == nil {
		return nil
	}
	if err := c.bus.EnsureConsumerGroup(ctx); err != nil {
		if c.logger != nil {
			c.logger.Error("notification consumer group setup failed", "error", err)
		}
		return err
	}
	go c.loop(ctx)
	if c.logger != nil {
		c.logger.Info("notification consumer started")
	}
	return nil
}

func (c *Consumer) Stop(ctx context.Context) error {
	close(c.stop)
	select {
	case <-c.done:
	case <-ctx.Done():
	}
	return nil
}

func (c *Consumer) loop(ctx context.Context) {
	defer close(c.done)
	for {
		select {
		case <-c.stop:
			return
		case <-ctx.Done():
			return
		default:
		}
		messages, err := c.bus.ReadGroup(ctx, c.name, 20)
		if err != nil {
			if c.logger != nil {
				c.logger.Warn("notification consumer read failed", "error", err)
			}
			time.Sleep(2 * time.Second)
			continue
		}
		for _, msg := range messages {
			var createdAt *time.Time
			if msg.CreatedAt != "" {
				if t, err := time.Parse(time.RFC3339, msg.CreatedAt); err == nil {
					createdAt = &t
				}
			}
			if _, err := c.repo.UpsertFromBus(ctx, msg.Fingerprint, msg.PortfolioID, msg.Kind, msg.Payload, msg.Urgency, createdAt); err != nil {
				if c.logger != nil {
					c.logger.Warn("notification consumer upsert failed", "error", err)
				}
				continue
			}
			_ = c.bus.Ack(ctx, msg.MessageID)
		}
	}
}

var _ application.NotificationConsumer = (*Consumer)(nil)
